livre/lib/data/BoundingAxis.cpp

```diff
@@ -0,0 +1,374 @@
+                normal[ axis ] = 0.0;
+
+                vertices.push_back( vertex );
+                normals.push_back( normal );
+
+                if( k != 0 )
+                {
+                    vertices.push_back( vertex );
+                    normals.push_back( normal );
+                }
+
+                if( k == 3 )
+                {
+                    vertices.push_back( vertices[ vertices.size() - 7 ]);
+                    normals.push_back( normals[ normals.size() - 7 ]);
+                }
+            }
+        }
+    }
+
+    for( size_t i = 0; i < vertices.size(); ++i )
+    {
+        verticesBuffer.push_back( vertices[ i ].x( ));
+        verticesBuffer.push_back( vertices[ i ].y( ));
+        verticesBuffer.push_back( vertices[ i ].z( ));
+
+        normalsBuffer.push_back( normals[ i ].x( ));
+        normalsBuffer.push_back( normals[ i ].y( ));
+        normalsBuffer.push_back( normals[ i ].z( ));
+        normalsBuffer2.push_back( normals[ i ].x( ));
+        normalsBuffer2.push_back( normals[ i ].y( ));
+        normalsBuffer2.push_back( normals[ i ].z( ));
+
+        colorsBuffer.push_back( color.r( ));
+        colorsBuffer.push_back( color.g( ));
+        colorsBuffer.push_back( color.b( ));
+        colorsBuffer.push_back( color.a( ));
+
+        typesBuffer.push_back( BBOX );
+
+        ++nVertices;
+    }
+
+    _createTicks( volInfo, verticesBuffer, normalsBuffer, normalsBuffer2, colorsBuffer,
+                  typesBuffer );
+
+    _sendDataToGPU( verticesBuffer, normalsBuffer, normalsBuffer2, colorsBuffer, typesBuffer );
+}
+
+BoundingAxis::~BoundingAxis()
+{
+    _cleanGPUMemory();
+}
+
+void BoundingAxis::_sendDataToGPU( const Floats& vertices, const Floats& normals,
+                                   const Floats& normals2, const Floats& colors,
+                                   const UInt8s& types )
+{
+    glGenBuffers( 1, &_vertexeBufferId );
+    glBindBuffer( GL_ARRAY_BUFFER, _vertexeBufferId );
+    glBufferData( GL_ARRAY_BUFFER, sizeof(float) * vertices.size(),
+                  &vertices[0], GL_STATIC_DRAW );
+
+    glGenBuffers( 1, &_normalBufferId );
+    glBindBuffer( GL_ARRAY_BUFFER, _normalBufferId );
+    glBufferData( GL_ARRAY_BUFFER, sizeof(GLuint) * normals.size(),
+                  &normals[0], GL_STATIC_DRAW );
+
+    glGenBuffers( 1, &_normalBufferId2 );
+    glBindBuffer( GL_ARRAY_BUFFER, _normalBufferId2 );
+    glBufferData( GL_ARRAY_BUFFER, sizeof(GLuint) * normals2.size(),
+                  &normals2[0], GL_STATIC_DRAW );
+
+    glGenBuffers( 1, &_colorBufferId );
+    glBindBuffer( GL_ARRAY_BUFFER, _colorBufferId );
+    glBufferData( GL_ARRAY_BUFFER, sizeof(GLuint) * colors.size(),
+                  &colors[0], GL_STATIC_DRAW );
+
+    glGenBuffers( 1, &_typeBufferId );
+    glBindBuffer( GL_ARRAY_BUFFER, _typeBufferId );
+    glBufferData( GL_ARRAY_BUFFER, sizeof(GLubyte) * types.size(),
+                  &types[0], GL_STATIC_DRAW );
```

Comment:
Normally, a single VBO is used for all the data so you don't have to rebind a new buffer to set each vertex attrib pointer. You can use glBufferSubData to upload the pieces one by one.